File: tools/editor/editor/src/com/lalalic/editor/view/Selection.java
package com.lalalic.editor.view;

import com.google.gwt.canvas.client.Canvas;
import com.google.gwt.canvas.dom.client.Context2d;
import com.google.gwt.user.client.Timer;

public class Selection extends View{
	public static final int INTERVAL=500;
	private Editor editor;
	public Location start=new Location(),end=new Location();
	private Timer timer;
	private Rect last=new Rect();
	private Context2d paint;
	public Selection(Editor editor){
		this.editor=editor;
	}

	public void onDraw(final Canvas canvas){
		if(timer==null){
			this.paint=canvas.getContext2d();
			start.x=canvas.getCanvasElement().getAbsoluteLeft();
			start.y=canvas.getCanvasElement().getAbsoluteTop();
			timer=new Timer(){
				@Override
				public void run() {
					double height=editor.getMetrics().getHeight();
					if(last.left!=-1){
						paint.clearRect(last.left,last.top, last.right,last.bottom);
						last.clear();
					}else{
						paint.beginPath();
						double w=paint.getLineWidth();
						paint.moveTo(start.x+w/2, start.y);
						paint.lineTo(start.x+w/2, start.y+height);
						last.set(start.x, start.y, start.x+w, start.y+height);
						paint.stroke();
					}
				}
			};
		}
		validate();
	}
	
	public void setStart(int start) {
		this.start.doc = start;
	}

	public Location getStart() {
		return start;
	}

	public void setEnd(int end) {
		this.end.doc = end;
	}

	public Location getEnd() {
		return end;
	}
	
	public void invalid(){
		if(last.left!=-1){
			paint.clearRect(last.left,last.top, last.right,last.bottom);
			last.clear();
		}
		this.timer.cancel();
	}
	
	public void validate(){
		this.timer.scheduleRepeating(INTERVAL);
	}

	public class Location{
		int doc;
		int line;
		int character;
		double x,y;
	}
	
	public class Rect{
		double left=-1, top=-1, right=-1,bottom=-1;
		public void set(double left, double top, double right, double bottom){
			this.left=left;
			this.top=top;
			this.right=right;
			this.bottom=bottom;
		}
		public void clear(){
			left=-1;
			top=-1; 
			right=-1;
			bottom=-1;
		}
	}
	
	
}

File: tools/editor/js/hua/view/Selection.js
define(['view/View'],function(View){
	var Selection=function(){
			View.call(this,arguments);
		}, input;
	Selection.prototype=new View();
	Selection.prototype.INTERVAL=500;
	Selection.prototype.start=Selection.prototype.end=null
	Selection.prototype.last={left:-1}
	View.prototype.clazz="Selection"
	function _draw(canvas){
		if(!this.start){
			this.end=this.start={
				x:input.position.left+input.getContentLeft(), 
				y:input.position.top+input.getContentTop()
			};
		}
		var paint=canvas.getContext("2d"),
			w=paint.lineWidth;
			height=input.metrics.height;
		if(this.last.left==-1){
			this.last={left:this.start.x, top:this.start.y, right:this.start.x+w, bottom:this.start.y+height};
			paint.beginPath();
			paint.moveTo(Math.floor(this.start.x+w/2)+0.5, this.start.y);
			paint.lineTo(Math.floor(this.start.x+w/2)+0.5, this.start.y+height);
			paint.stroke();
		}else{
			this.clean();
		}
	};
	Selection.prototype.setHolder=function(view){
		input=view
		this.position.left=input.position.left
		this.position.top=input.position.top
		this.end=this.start=null;
	};
	
	Selection.prototype.getHolder=function(){
		return input
	};
	
	Selection.prototype.onDraw=function(canvas){
		if(!this.timer){
			var me=this;
			this.timer=setInterval(function(){_draw.call(me,canvas);},500);
		}
	};
	Selection.prototype.clean=function(){
		this.paint.clearRect(this.last.left,this.last.top,this.last.right, this.last.bottom);
		this.last={left:-1};
	}
	Selection.prototype.validate=function(){
		this.draw(this.canvas);
	};
	Selection.prototype.invalidate=function(){
		if(this.timer){
			clearInterval(this.timer);
			this.clean();
			this.timer=null;			
		}
	};
	Selection.prototype.isRoot=function(){
		return false
	}
	return new Selection();
})